### ISSUE TITLE
ui-control possibile memory leak

### ISSUE THREAD

**Issue by gtnscebba** [FlowFuse/node-red-dashboard#1281]
### Current Behavior

I use ui-control on a subflow (named "Trigger-DB2"), 15 instances.
When a page is refreshed, on syslog appear these messages:

```
2024-09-10T09:09:27.684669+02:00 PI4bOpenVPN Node-RED[502]: (node:502) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 disconnect listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
2024-09-10T09:09:27.699870+02:00 PI4bOpenVPN Node-RED[502]: (node:502) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 ui-control listeners added to [Socket]. MaxListeners is 10. Use emitter.setMaxListeners() to increase limit
```

### Expected Behavior

If the problem is the low limit of MaxListeners, increase the number of them

### Steps To Reproduce

Deploy more than 10 intances of ui-control to trigger events when page refreshes, then refresh a page. 

### Environment

- Dashboard version: 1.16.0
- Node-RED version: 4.0.2
- Node.js version: 20.17.0
- npm version: 10.8.2
- Platform/OS: SMP Raspbian 1:6.6.47-1+rpt1
- Browser: Chrome 128.0.6613.120 (64 bit)

### Have you provided an initial effort estimate for this issue?

I am not a FlowFuse team member

**Comment by bartbutenaers** [comment]
Just adding the original [discussion](https://discourse.nodered.org/t/ui-control-possible-memory-leak/90768/4?u=bartbutenaers) , where @gtnscebba has provided a lot of extra information.

**Comment by bartbutenaers** [comment]
@gtnscebba 
The most interesting part (for me at least) is the line number where the exception occurs, but that is missing in your log:

![image](https://github.com/user-attachments/assets/bd5bcd52-df77-46ba-b08f-a97d58be98a4)

**Comment by gtnscebba** [comment]
```
2024-09-11T11:17:47.844282+02:00 PI4bOpenVPN Node-RED[502]: 11 Sep 11:17:47 - [red] Uncaught Exception:
2024-09-11T11:17:47.844745+02:00 PI4bOpenVPN Node-RED[502]: 11 Sep 11:17:47 - [error] TypeError: Cannot read properties of null (reading 'send')
2024-09-11T11:17:47.844890+02:00 PI4bOpenVPN Node-RED[502]:     at ui-control (/home/pi/.node-red/node_modules/@flowfuse/node-red-dashboard/nodes/widgets/ui_control.js:186:35)
2024-09-11T11:17:47.844996+02:00 PI4bOpenVPN Node-RED[502]:     at Socket.emit (node:events:531:35)
2024-09-11T11:17:47.845095+02:00 PI4bOpenVPN Node-RED[502]:     at Socket.emitUntyped (/home/pi/.node-red/node_modules/socket.io/dist/typed-events.js:69:22)
2024-09-11T11:17:47.845204+02:00 PI4bOpenVPN Node-RED[502]:     at /home/pi/.node-red/node_modules/socket.io/dist/socket.js:704:39
2024-09-11T11:17:47.845305+02:00 PI4bOpenVPN Node-RED[502]:     at processTicksAndRejections (node:internal/process/task_queues:77:11)
2024-09-11T11:17:47.872252+02:00 PI4bOpenVPN systemd[1]: nodered.service: Main process exited, code=exited, status=1/FAILURE
2024-09-11T11:17:47.872888+02:00 PI4bOpenVPN systemd[1]: nodered.service: Failed with result 'exit-code'.
2024-09-11T11:17:47.873219+02:00 PI4bOpenVPN systemd[1]: nodered.service: Consumed 5h 23min 56.244s CPU time.
2024-09-11T11:18:08.111058+02:00 PI4bOpenVPN systemd[1]: nodered.service: Scheduled restart job, restart counter is at 1.
2024-09-11T11:18:08.114168+02:00 PI4bOpenVPN systemd[1]: Stopped nodered.service - Node-RED graphical event wiring tool.
2024-09-11T11:18:08.114687+02:00 PI4bOpenVPN systemd[1]: nodered.service: Consumed 5h 23min 56.244s CPU time.
2024-09-11T11:18:08.181793+02:00 PI4bOpenVPN systemd[1]: Started nodered.service - Node-RED graphical event wiring tool.
2024-09-11T11:18:10.184911+02:00 PI4bOpenVPN Node-RED[11172]: 11 Sep 11:18:10 - [info]
2024-09-11T11:18:10.185358+02:00 PI4bOpenVPN Node-RED[11172]: Welcome to Node-RED

```

I have enlarged the window. The row is ui_control.js:186:35

**Comment by bartbutenaers** [comment]
Hi @gtnscebba,

I 'think' multiple separate issues are being reported here.

## Issue 1 - Uncaught Exception
I cannot really explain your error.  It occurs [here](https://github.com/FlowFuse/node-red-dashboard/blob/main/nodes/widgets/ui_control.js#L186) in the code on line 186:

![image](https://github.com/user-attachments/assets/68ca8c00-f654-4bf0-b702-a4505da7058c)

It looks like some event arrives (via a socketio websocket) from the client.  It tries to find the ui-control node with that id, but that seems not exist anymore so it fails.

+ I don't think this has to do anything with a memory leak.
+ I don't use the ui-control node myself, so not familar with it.  You said you called the dashboard in the morning.  Could it be that the ui-control node had been removed meanwhile from the flow?  Not sure why else this could occur.
+ The undefined wNode should be detected in the code and handled gracefully instead of crashing.
+ Not sure why the code contains `RED.nodes.getNode(node.id)`, because a first sight it looks to me that it would (in a happy path) simply `node` which we have already...

## Issue 2 - MaxListenersExceededWarning
This warning occurs when you have more than 10 event listeners for the ***same*** event.   When you have mutlple event handlers without removing them, the memory usage would increase (resulting in a memory leak).  Therefore the default limit is 10 event listeners.

 I see on your discourse discussion that you have 15 identical subflows containing a ui-control node.   Could that perhaps cause this warning?  Can you perhaps try experimenting with the number of ui-control nodes in your flows.  Perhaps the warning disappears when you have in total less than 10 ui-control nodes?  Don't know...

**Comment by gtnscebba** [comment]
Thank you for your last message.
Since it was very complicated for me to solve this problem (without going deeply into the code), I decided to radically change the approach and now the error on the syslog does not appear anymore.

As I said in my first post, I used ui-control inside a subflow that was called every time I had to change page and start a flow.

So I decided to use a single instance of ui-control and send the message of the changed page to all the parts of the code involved (those in which the removed subflow was present).
Here, with a switch node it is tested if msg.name contains the name of the part involved at that moment. If so, the okay is given to continue processing; otherwise nothing is triggered.
With this way of operating on the syslog only the informative message of close appears when the page is refreshed

`2024-09-15T07:24:37.707327+02:00 PI4bOpenVPN Node-RED[12375]: 15 Sep 07:24:37 - [info] [ui-base:My Dashboard] Disconnected GYDnjYePUOpG1zdcAACo due to transport close
`

Basically, this is what I did:

![image](https://github.com/user-attachments/assets/b5480bd0-4e0f-4f25-8631-776444481dbb)

In the flow:

![image](https://github.com/user-attachments/assets/3584b326-d158-42af-8d0f-ca2f062a094b)

I don't know if this is the best way to do things, but it certainly solved my problem.

**Comment by bartbutenaers** [comment]
Hi @gtnscebba ,
Thanks for sharing your workaround in detail, because that might help others running into the same problem!

If anybody else ever arrives here by searching for _**MaxListenersExceededWarning**_, it might be useful to experiment whether less than 10 ui_control nodes in the flow solves the problem.  In that case I assume there are 2 possible solutions, to solve this problem forever in the dashboard codebase:
1. Increase the maximum number of listeners.  I think that is not a problem, because in the cleanup code of the ui-control node, the listener will be automatically cleaned up.  So no memory leak introduced as the warning suggests (that it might happen if you don't cleanup the listeners).

2. Refactor the code of the ui-control node.  So that you have only one listener, and pass the info the the ui-control node instance for which it is intended.  Not sure if that is possible, because I had no time to look at the ui-control node code in depth.

Personally I would prefer option 2...
